### PR TITLE
Support for custom icons

### PR DESCRIPTION
**Pull request by patosai**
#11.

out

### CODE REVIEW

**Comment by gsingh93** [comment]
This looks awesome! Thanks for the PR. It'll take us some time to get to it, we'll do our best.

**Comment by patosai** [comment]
Thanks for hosting the Michigan Hackers CTF, I thoroughly enjoyed it. And you carried me through "baby's first bof."

**Comment by colbyjoe1919** [comment]
Yeah work sucks.  But you know what really tanks it's that whole afterlife
in hell .  You guys are fine though unless someone informs you that you are
pissing off God then it's too late.  I dang I just did my bad.

**Comment by javuto** [comment]
@patosai this is amazing! Like @gsingh93 said, it will take us a minute to go through the code and test it but looks great. Good job!

**Comment by kazenomamani210** [comment]
I can not see this option on team registation... does that mean it is not avarable?

**Comment by javuto** [comment]
@patosai can you rebase with the latest changes and see if we can push this through? Thank you!

**Comment by patosai** [comment]
@javuto check it out

**Comment by javuto** [comment]
Sorry @patosai, we did push master with changes to the provision script so you may need to rebase again. Thank you!

File: src/controllers/IndexController.php [hunk 1]
```diff
@@ -389,8 +405,29 @@ class="fb-main page--registration full-height fb-scroll">
               {$token_field}
             </fieldset>
             <div class="fb-choose-emblem">
+<<<<<<< HEAD
```
Comment (iangcarroll) [inline src/controllers/IndexController.php]: Looks like a merge left something behind. :smile:

Comment (patosai) [inline src/controllers/IndexController.php]: Whoops! I'm going to undo the previous merge and merge again then.

**Comment by patosai** [comment]
@javuto working on my machine!

**Comment by kazenomamani210** [comment]
When is this function avarable?

**Comment by javuto** [comment]
Going to focus on this to get it merged as soon as possible. @patosai can you make sure you have rebased with the latest changes? Thank you!

**Comment by javuto** [comment]
Tried to use this code as it is, and noticed few problems. The first one is that I added the method ```Logo:::genCreate``` for the import functionality and that conflicts with your ```genCreate```. Also the file chooser does not seem to work properly and the request fails, even with a regular logo. Here the errors:
```
[Thu Sep 15 01:56:00 2016] [hphp] [14140:7f97683ff700:32:000001] [] \nFatal error: Uncaught exception 'HH\\InvariantException' with message 'Expected logo to be a string' in /var/www/fbctf/src/Utils.php:22\nStack trace:\n#0 /var/www/fbctf/src/Utils.php(22): HH\\invariant_violation()\n#1 /var/www/fbctf/src/controllers/ajax/IndexAjaxController.php(55): must_have_string()\n#2 /var/www/fbctf/src/controllers/ajax/AjaxController.php(14): IndexAjaxController->genHandleAction()\n#3 /var/www/fbctf/src/Router.php(52): AjaxController->genHandleRequest()\n#4 /var/www/fbctf/src/Router.php(14): Router::genRouteAjax()\n#5 /var/www/fbctf/src/index.php(7): Router::genRoute()\n#6 /var/www/fbctf/src/index.php(17): genInit()\n#7 {main}
[Thu Sep 15 01:56:13 2016] [hphp] [14140:7f97683ff700:33:000001] [] \nFatal error: Uncaught exception 'HH\\InvariantException' with message 'Index isCustomLogo not found in container' in /var/www/fbctf/src/Utils.php:13\nStack trace:\n#0 /var/www/fbctf/src/Utils.php(13): HH\\invariant_violation()\n#1 /var/www/fbctf/src/Utils.php(39): must_have_idx()\n#2 /var/www/fbctf/src/controllers/ajax/IndexAjaxController.php(55): must_have_bool()\n#3 /var/www/fbctf/src/controllers/ajax/AjaxController.php(14): IndexAjaxController->genHandleAction()\n#4 /var/www/fbctf/src/Router.php(52): AjaxController->genHandleRequest()\n#5 /var/www/fbctf/src/Router.php(14): Router::genRouteAjax()\n#6 /var/www/fbctf/src/index.php(7): Router::genRoute()\n#7 /var/www/fbctf/src/index.php(17): genInit()\n#8 {main}
```

**Comment by patosai** [comment]
Everything should work now.

**Comment by patosai** [comment]
..except Travis is complaining, seemingly about Unison.

**Comment by javuto** [comment]
Don't worry about the build. We can re-run it before merge.

File: .gitignore [hunk 2]
```diff
@@ -34,3 +34,7 @@ src/data/attachments/*
 !src/data/attachments/index.php
 src/data/attachments/deleted/*
 !src/data/attachments/deleted/index.php
+
+# Custom logos directory
```
Comment (javuto) [inline .gitignore]: I think is better if you make the custom icons directory in ```src/data/customlogos``` and make sure you create that directory in the provision script and make it world writable: https://github.com/facebook/fbctf/blob/master/extra/provision.sh#L252

File: database/logos.sql [hunk 3]
```diff
@@ -10,6 +10,7 @@ CREATE TABLE `logos` (
   `used` tinyint(1) DEFAULT 1,
   `enabled` tinyint(1) DEFAULT 1,
   `protected` tinyint(1) DEFAULT 0,
+  `custom` tinyint(1) DEFAULT 1,
```
Comment (javuto) [inline database/logos.sql]: The default ```custom``` should be ```0```, I believe

File: database/schema.sql [hunk 4]
```diff
@@ -184,7 +184,8 @@ CREATE TABLE `configuration` (
 LOCK TABLES `configuration` WRITE;
 INSERT INTO `configuration` (field, value, description) VALUES("game", "0", "(Boolean) Game is ongoing");
 INSERT INTO `configuration` (field, value, description) VALUES("next_game", "0", "(Date) Next game to happen");
-INSERT INTO `configuration` (field, value, description) VALUES("game_duration", "10800", "(Integer) Duration of game in seconds");
+INSERT INTO `configuration` (field, value, description) VALUES("game_duration_value", "3", "(Integer) Value of the duration of the game");
```
Comment (javuto) [inline database/schema.sql]: Doesn't this code belong to a different commit that was just landed this week? The one for the custom duration: https://github.com/facebook/fbctf/commit/bd8db7ab8b560331898a19fe5bae463cabea72b0

File: .gitignore (hunk 2 above)
Comment (javuto) [inline .gitignore]: Oh nevermind the comment about the provision, just saw it 👍

File: extra/provision.sh [hunk 5]
```diff
@@ -250,6 +250,8 @@ import_empty_db "root" "$P_ROOT" "$DB" "$CTF_PATH" "$MODE"
 # Make attachments folder world writable
 sudo chmod 777 "$CTF_PATH/src/data/attachments"
 sudo chmod 777 "$CTF_PATH/src/data/attachments/deleted"
+# Make custom logos folder world writable
+sudo chmod 777 "$CTF_PATH/src/static/img/customlogo"
```
Comment (javuto) [inline extra/provision.sh]: See my comment above to make this ```src/data/customlogos```

File: src/controllers/AdminController.php [hunk 6]
```diff
@@ -150,24 +150,28 @@ class="fb--conf--registration_type"
   }
 
   private async function genConfigurationDurationSelect(): Awaitable<:xhp> {
-    $config = await Configuration::gen('game_duration');
-    $duration = intval($config->getValue());
-    $select = <select name="fb--conf--game_duration"></select>;
+    $config = await Configuration::gen('game_duration_unit');
```
Comment (javuto) [inline src/controllers/AdminController.php]: Again, this belongs to a different commit.

File: database/schema.sql (hunk 4 above)
Comment (patosai) [inline database/schema.sql]: It does, and I believe I had a merge conflict on that line and I changed it to the current line. Perhaps I should have left the old one?

**Comment by javuto** [comment]
In fact, there are two things that we should check:
- Make sure that the file is an image: Extension, mimetype or headers.
- Apply a size limit to the image. Otherwise it will impact the overall performance.

**Comment by gsingh93** [comment]
Not much he can learn from the attachments code for this. But the code needs to first check whether the file is a valid PNG/JPG, and then change the extension to .png or .jpg.

**Comment by gsingh93** [comment]
Also, the only file type you can trust is getimagesize(). Do *not* trust the mimetype.

File: src/controllers/ajax/IndexAjaxController.php [hunk 7]
```diff
@@ -119,6 +131,25 @@ protected function getActions(): array<string> {
...
+
+    if ($is_custom_logo) {
+      $logo_data = str_replace(' ', '+', $final_logo);
+      $logo_data_unbase64ed = base64_decode($logo_data);
+
+      $name = 'custom-'.time();
+      $filename = $name.'.'.$logo_type;
+      // '..' is separated for DB storage
+      $url = self::CUSTOM_LOGO_DIR.$filename;
+      $filepath = dirname(__FILE__).'/../..'.$url;
```
Comment (javuto) [inline src/controllers/ajax/IndexAjaxController.php]: I would prefer avoiding the "../../" type of expressions. The path for custom logos is always the same, and since you are generating the filename, just use them statically.

File: src/controllers/ajax/IndexAjaxController.php [hunk 8]
```diff
@@ -119,6 +131,25 @@ protected function getActions(): array<string> {
 
     // Check logo
     $final_logo = $logo;
+
+    if ($is_custom_logo) {
+      $logo_data = str_replace(' ', '+', $final_logo);
+      $logo_data_unbase64ed = base64_decode($logo_data);
+
+      $name = 'custom-'.time();
```
Comment (javuto) [inline src/controllers/ajax/IndexAjaxController.php]: A bit more of entropy here would be nice, just the timestamp won't do it. maybe md5 the filename and get a chunk of it?

File: src/controllers/ajax/IndexAjaxController.php [hunk 9]
```diff
@@ -119,6 +131,25 @@ protected function getActions(): array<string> {
...
+      $filename = $name.'.'.$logo_type;
+      // '..' is separated for DB storage
+      $url = self::CUSTOM_LOGO_DIR.$filename;
+      $filepath = dirname(__FILE__).'/../..'.$url;
+      file_put_contents($filepath, $logo_data_unbase64ed);
+
+      // Check validity of logo file
+      list($width, $height, $type, $attr) = getimagesize($full_filename);
```
Comment (javuto) [inline src/controllers/ajax/IndexAjaxController.php]: This check should be done before putting the file in the final destination. Also here is where you want to make sure the uploaded file is not a ```.php``` file. Easiest way would be to limit the extensions to png, gif and jpg.

**Comment by javuto** [comment]
The build is failing because can not find the ```customlogos``` folder. You can add a ```mkdir -p "$CTF_PATH/src/data/customlogos"``` somewhere around line 250 in ```provision.sh``` and the build will work.

**Comment by patosai** [comment]
I'll get back to this this weekend; went to HackMIT so catching up on homework.

**Comment by patosai** [comment]
The admin panel shows the logos now, and the images are checked before written to disk. I tried uploading all sorts of junk and none of it got through.

**Comment by kazenomamani210** [comment]
@javuto Any progress here?

**Comment by juliannagler** [comment]
Hey @kazenomamani210 we'll do a final review of this in the next couple weeks and merge if everything looks good. At the moment we're focused on the needed sessions work.

**Comment by patosai** [comment]
Looks like this can be closed now?

**Comment by javuto** [comment]
I believe so, the functionality was merged based on this one. Thanks a lot for the initial and hard work @patosai
